editor/metaboxes/index.js

```diff
@@ -0,0 +1,212 @@
+		this.dirtyObserver = new window.MutationObserver( this.checkMetaboxState );
+		this.dirtyObserver.observe( findDOMNode( form ), {
+			attributes: true,
+			attributeOldValue: true,
+			characterData: true,
+			characterDataOldValue: true,
+			childList: true,
+			subtree: true,
+		} );
+		form.addEventListener( 'change', this.checkMetaboxState );
+		form.addEventListener( 'input', this.checkMetaboxState );
+
+		this.checkMetaboxState();
+	}
+
+	getFormData( node ) {
+		if ( ! this.isFrameAccessible ) {
+			return;
+		}
+
+		const form = node.contentWindow.document.getElementById( 'post' );
+
+		const data = new window.FormData( form );
+		const entries = Array.from( data.entries() );
+		return entries;
+	}
+
+	checkMetaboxState() {
+		const entries = this.getFormData( this.node );
+
+		if ( ! isEqual( this.originalFormData, entries ) ) {
+			if ( this.props.metaboxes[ this.props.location ].isDirty === false ) {
+				this.props.metaboxStateChanged( this.props.location, true );
+			}
+
+			return;
+		}
+
+		// If the data is the same as the original and we have metabox marked as dirty.
+		if ( this.props.metaboxes[ this.props.location ].isDirty === true ) {
+			this.props.metaboxStateChanged( this.props.location, false );
+		}
+	}
+
+	handleMetaboxReload( event ) {
+		event.target.removeEventListener( 'load', this.handleMetaboxReload );
+		this.props.metaboxReloaded( this.props.location );
+	}
+
+	checkMessageForResize( event ) {
+		const iframe = this.node;
+
+		// Attempt to parse the message data as JSON if passed as string
+		let data = event.data || {};
+		if ( 'string' === typeof data ) {
+			try {
+				data = JSON.parse( data );
+			} catch ( e ) {} // eslint-disable-line no-empty
+		}
+
+		if ( data.source !== 'metabox' || data.location !== this.props.location ) {
+			return;
+		}
+
+		// Verify that the mounted element is the source of the message
+		if ( ! iframe || iframe.contentWindow !== event.source ) {
+			return;
+		}
+
+		// Update the state only if the message is formatted as we expect, i.e.
+		// as an object with a 'resize' action, width, and height
+		const { action, width, height } = data;
+		const { width: oldWidth, height: oldHeight } = this.state;
+
+		if ( 'resize' === action && ( oldWidth !== width || oldHeight !== height ) ) {
+			this.setState( { width, height } );
+		}
+	}
+
+	render() {
+		const { location, isSidebarOpened, className, id = 'gutenberg-metabox-iframe' } = this.props;
+		const classes = classnames( {
+			className,
+			'sidebar-open': isSidebarOpened,
+		} );
+
+		return (
+			<iframe
+				ref={ ( node ) => {
+					this.node = node;
+				} }
+				title={ __( 'Extended Settings' ) }
+				key="metabox"
+				id={ id }
+				className={ classes }
+				src={ `${ window._wpMetaboxUrl }&metabox=${ location }` }
+				width={ Math.ceil( this.state.width ) }
+				height={ Math.ceil( this.state.height ) } />
+		);
+	}
+}
+
+function mapStateToProps( state ) {
+	return {
+		metaboxes: state.legacyMetaboxes,
```

Comment:
maybe a selector?

Reply:
Also, should we be more specific here?: `isMetaboxDirty( state, location )` and `isMetaboxLoading( state, location )`

Reply:
Yeah, that sounds like a good idea.